src/util/performance.cpp

```diff
@@ -103,40 +133,128 @@ void end_time_block(u64 hash_uuid) {
     TimePoint start = ref.start;
     ref.start = NULL_TIME;
     ref.total_nano_seconds += time_since(start, end);
+
+    record_to_performance_capture_file('E', ref.name, ref.func, ref.file, ref.line);
+}
+
+//
+// Functions related to performance capture
+//
+// State for the capture, shouldn't be neeeded
+// outside of this code.
+namespace Capture {
+    const char CAPTURE_FILE_NAME[] = "perf-capture.json";
+    i32 frames_to_capture = 0;
+}
+
+void capture_handle() {
+    if (Capture::frames_to_capture != 0 && !capture_file) {
+        LOCK_FOR_BLOCK(capture_file_mutex);
+        capture_file = fopen(Capture::CAPTURE_FILE_NAME, "w");
+        const char preamble[] = "[{}";
+        fwrite((void *)preamble, 1, LEN(preamble) - 1, capture_file);
+    }
+
+    if (Capture::frames_to_capture != 0) {
+        if (Capture::frames_to_capture > 0)
+            Capture::frames_to_capture--;
+
+        if (Capture::frames_to_capture == 0) {
+            LOCK_FOR_BLOCK(capture_file_mutex);
+            const char postamble[] = "]";
+            fwrite((void *)postamble, 1, LEN(postamble) - 1, capture_file);
+            fclose(capture_file);
+            capture_file = nullptr;
+        }
+    }
+}
+
+void capture_begin() {
+    // Capture indefinately.
+    Capture::frames_to_capture = -1;
+}
+
+void capture_end() {
+    // Capture until next frame.
+    Capture::frames_to_capture = 1;
+}
+
+void capture_frames(i32 n) {
+    ASSERT_NE(n, 0);
+    Capture::frames_to_capture = Math::max(0, Capture::frames_to_capture) + 100;
+}
+
+bool capture_is_running() {
+    return capture_frames_left() != 0;
+}
+
+i32 capture_frames_left() {
+    return Capture::frames_to_capture;
 }
 
 #ifdef IMGUI_ENABLE
 const f32 NANO_TO_MS = 1e-6;
 
 TimePoint frame_start = {};
 f32 frame_time[HISTORY_LENGTH] = {};
-u32 frame = 0;
+
+// TODO(ed):
+// Break this function into sub-functions
 
 void report() {
+
+    TimePoint now = Clock::now();
+    if (frame_start.time_since_epoch().count() == 0) {
+        frame_start = now;
+    }
+    u32 frame_index = frame() % HISTORY_LENGTH;
+
+    f32 current_frame_time = time_since(frame_start, now) * NANO_TO_MS;
+    frame_time[frame_index] = current_frame_time;
+    frame_start = now;
+
+    record_to_performance_capture_file('i', "FRAME", "NA", "NA", 0);
+    capture_handle();
+
     if (!GAMESTATE()->imgui.performance_enabled) return;
     ImGui::Begin("Performance");
 
+    if (ImGui::Button("Start Capture")) {
+        capture_begin();
+    }
+    if (capture_is_running()) {
+        ImGui::SameLine();
+        if (ImGui::Button("+100 frames")) {
+            capture_frames(100);
+        }
```

Comment:
I guess you do a `ImGui::InputScalar` here. It takes an input_step parameter so we can step the capture size in steps of e.g. 10, which is nice.

Reply:
Why not a `ImGui::InputInt`? ;)

Reply:
Yes that's true, I don't know how I only found scalar in the web manual.